npm/vite-dev-server-fresh/src/plugins/cypress.ts

```diff
@@ -0,0 +1,157 @@
+import debugFn from 'debug'
+import { resolve } from 'pathe'
+import type { ModuleNode, Plugin, ViteDevServer } from 'vite'
+import { normalizePath } from 'vite'
+import fs from 'fs'
+
+import type { ViteDevServerConfig } from '../devServer'
+import path from 'path'
+
+const debug = debugFn('cypress:vite-dev-server:plugins:cypress')
+
+const INIT_FILEPATH = resolve(__dirname, '../../client/initCypressTests.js')
+
+const HMR_DEPENDENCY_LOOKUP_MAX_ITERATION = 50
+
+interface Spec {
+  absolute: string
+  relative: string
+}
+
+function getSpecsPathsSet (specs: Spec[]) {
+  return new Set<string>(
+    specs.map((spec) => spec.absolute),
+  )
+}
+
+export const Cypress = (
+  options: ViteDevServerConfig,
+): Plugin => {
+  let base = '/'
+
+  const projectRoot = options.cypressConfig.projectRoot
+  const supportFilePath = options.cypressConfig.supportFile ? path.resolve(projectRoot, options.cypressConfig.supportFile) : false
+
+  const devServerEvents = options.devServerEvents
+  const specs = options.specs
+  const indexHtmlFile = options.cypressConfig.indexHtmlFile
+
+  let specsPathsSet = getSpecsPathsSet(specs)
+  let loader = fs.readFileSync(INIT_FILEPATH, 'utf8')
+
+  devServerEvents.on('dev-server:specs:changed', (specs: Spec[]) => {
+    specsPathsSet = getSpecsPathsSet(specs)
+  })
+
+  return {
+    name: 'cypress:main',
+    enforce: 'pre',
+    configResolved (config) {
+      base = config.base
+    },
+    async transformIndexHtml () {
+      const indexHtmlPath = resolve(projectRoot, indexHtmlFile)
+
+      debug('resolved the indexHtmlPath as', indexHtmlPath, 'from', indexHtmlFile)
+      const indexHtmlContent = await fs.promises.readFile(indexHtmlPath, { encoding: 'utf8' })
+      // find </body> last index
+      const endOfBody = indexHtmlContent.lastIndexOf('</body>')
+
+      // insert the script in the end of the body
+      return `${indexHtmlContent.substring(0, endOfBody)
+    }<script>
+    ${loader}
+    </script>${
+      indexHtmlContent.substring(endOfBody)
+    }`
+    },
+    configureServer: async (server: ViteDevServer) => {
+      server.middlewares.use(`${base}index.html`, async (req, res) => {
+        let transformedIndexHtml = await server.transformIndexHtml(base, '')
+        const viteImport = `<script type="module" src="${options.cypressConfig.devServerPublicPathRoute}/@vite/client"></script>`
+
+        // If we're doing cy-in-cy, we need to be able to access the Cypress instance from the parent frame.
+        if (process.env.CYPRESS_INTERNAL_VITE_OPEN_MODE_TESTING) {
```

Comment:
This change was needed to get cypress-in-cypress working

Reply:
this shouldn't be necessary - the reason this is happening is because `top` is setting `document.domain` for `e2e` mode. If top stops doing that for component testing, its unnecessary to inject `document.domain` into the iframe.

This likely applies to webpack too - or any other situation where top needs to communicate with the iframe.

Reply:
if you have an frame that sets document.domain = '...' and even if you embed an iframe on the same domain it also needs to set that. If neither set it (and are on the same domain) they do not need to set document.domain to talk to one another.

Reply:
Got it, makes sense